data-science-onramp/data-ingestion/setup.py

```diff
@@ -0,0 +1,149 @@
+from random import choice, choices, randint, seed
+import sys
+
+from time import time_ns
+
+from google.cloud import bigquery
+
+from py4j.protocol import Py4JJavaError
+from pyspark.sql import SparkSession
+
+from pyspark.sql.functions import UserDefinedFunction
+from pyspark.sql.types import IntegerType, StringType
+
+
+# Create a SparkSession under the name "setup". Viewable via the Spark UI
+spark = SparkSession.builder.appName("setup").getOrCreate()
+
+bucket_name = sys.argv[1]
+upload = True  # Whether to upload data to BigQuery
+
+# Check whether or not results should be uploaded
+try:
+    sys.argv[2]
+    upload = False
+except IndexError:
+    print("Results will be uploaded to BigQuery")
+
+table = "bigquery-public-data.new_york_citibike.citibike_trips"
+
+# Check if table exists
+try:
+    df = spark.read.format('bigquery').option('table', table).load()
+except Py4JJavaError:
+    print(f"{table} does not exist. ")
+    sys.exit(0)
+
+# START MAKING DATA DIRTY
+
+
+def random_select(items, cum_weights):
+    '''Picks an item according to the cumulative weights'''
+    return choices(items, cum_weights=cum_weights, k=1)[0]
+
+
+def tripduration(duration):
+    '''Converts trip duration to other units'''
+    seconds = str(duration) + " s"
+    minutes = str(float(duration) / 60) + " min"
+    hours = str(float(duration) / 3600) + " h"
+    return random_select([seconds, minutes, hours, str(randint(-1000, -1))],
+                         [0.3, 0.6, 0.9, 1])
+
+
+def station_name(name):
+    '''Replaces '&' with '/' with a 50% chance'''
+    return choice([name, name.replace("&", "/")])
+
+
+def usertype(user):
+    '''Manipulates the user type string'''
+    return choice([user, user.upper(), user.lower(),
+                  "sub" if user == "Subscriber" else user,
+                   "cust" if user == "Customer" else user])
+
+
+def gender(s):
+    '''Manipulates the gender string'''
+    return choice([s, s.upper(), s.lower(),
+                  s[0] if len(s) > 0 else "",
+                   s[0].lower() if len(s) > 0 else ""])
+
+
+def convertAngle(angle):
+    '''Converts long and lat to DMS notation'''
+    degrees = int(angle)
+    minutes = int((angle - degrees) * 60)
+    seconds = int((angle - degrees - minutes/60) * 3600)
+    new_angle = str(degrees) + u"\u00B0" + \
+        str(minutes) + "'" + str(seconds) + '"'
+    return random_select([str(angle), new_angle], cum_weights=[0.55, 1])
+
+
+def dirty_data(proc_func, allow_none):
+    '''Master function returns a user defined function
+    that transforms the column data'''
+    def udf(col_value):
+        seed(hash(col_value) + time_ns())
+        if col_value is None:
+            return col_value
+        elif allow_none:
+            return random_select([None, proc_func(col_value)],
+                                 cum_weights=[0.05, 1])
+        else:
+            return proc_func(col_value)
+    return udf
```

Comment:
Does this require having a nested function?

Reply:
We have a nested function here because a User Defined Function can only take 1 argument (the column value). If we do not have this nested function, we will have repetitive code of calling the UDF for each column or have an unwieldy list comprehension. Which option do you  think is best?